Create DVL service

Williangalvani opened the pull request:
**This work has been pushed to [it's own branch](https://github.com/bluerobotics/companion/tree/dvl)**.
This Draft PR will stay here for now as there are important instructions to follow.

![image](https://user-images.githubusercontent.com/4013804/87191746-6a8d0d80-c2cb-11ea-9499-b6443c805188.png)


TO-DO:
 - [ ] testing
 - [ ] thread locks?
 - [x] save settings

to test, set the DVL ip address to 192.168.2.95, open a terminal in companion and do:

 - Start from companion 0.0.22
 - Make sure the ROV is connected to the internet
 - Navigate to http://192.168.2.2:2770/git
 - Click "origin" -> "branches" -> "dvl"
 - Click "update"
 - Wait until the update is over and the companion reboots
 - Refresh the page and "Water Linked" should show up in the navbar

Ardusub code is currently in [this branch](https://github.com/Williangalvani/ardupilot/tree/dvl_cleaned) (which needs some cleaning too) and a binary is available [~here~](https://github.com/bluerobotics/companion/files/4893717/ardusub.zip) [~here~](https://github.com/bluerobotics/companion/files/4949954/ardusub.zip) [merged to master](https://firmware.ardupilot.org/Sub/latest/Pixhawk1/ardusub.apj). 

suggested Parameters:

Use mavlink2!
```
SERIAL0_PROTOCOL 2
```
Enable odom
Use ekf3 for the odometry support
```
AHRS_EKF_TYPE 3
EK2_ENABLE 0
```
Enable odometry support
```
EK3_ENABLE 1
VISO_TYPE 1
```
Disable external gps in ekf3
```
EK3_GPS_TYPE 3
```
Suggested gains
```
PSC_POSXY_P 2.5
PSC_POSZ_P  1.0
PSC_VELXY_D 0.8
PSC_VELXY_I 0.5
PSC_VELXY_P 5.0
PSC_VELZ_P  5.0
```

patrickelectric commented on services/waterlinked/dvl-a50/mavlink2resthelper.py:
```diff
@@ -0,0 +1,125 @@
+        except:
+            return False
+
+        print(data)
```
Lost debug info ?

patrickelectric commented on services/waterlinked/dvl-a50/mavlink2resthelper.py:
```diff
@@ -0,0 +1,125 @@
+            result = requests.post(MAVLINK2REST_URL + '/mavlink', json=data)
+            return result.status_code == 200
+        except Exception as error:
+            report_status("error setting message frequency: " + str(error))
```
`Error setting` ?

patrickelectric commented on services/waterlinked/dvl-a50/mavlink2resthelper.py:
```diff
@@ -0,0 +1,125 @@
+    def send_vision(self, position_deltas, rotation_deltas=[0, 0, 0], confidence=100, dt=125000):
+        "Sends message VISION_POSITION_DELTA to flight controller"
+        # if not self.vision_template:
+        #     self.vision_template = json.loads(request(MAVLINK2REST_URL + '/helper/message/VISION_POSITION_DELTA'))
```
Lost development code ?

patrickelectric commented on services/waterlinked/dvl-a50/main.py:
```diff
@@ -0,0 +1,72 @@
+Driver for the Waterlinked DVL A-50
+"""
+
+from flask import Flask
```
missing setup.py to install flask ?

patrickelectric reviewed (changes requested):
Just a quick review, looks good in general

Williangalvani commented on br-webui/index.js:
```diff
@@ -49,14 +49,18 @@ app.all("/service/network*", function(req, res) {
  });
 
-app.all("/api/mavlink*", function(req, res) {
+ app.all("/api/mavlink*", function(req, res) {
```
a wild whitespace appears!
I'm doing a thorough review now to catch these smaller issues

Williangalvani commented on services/waterlinked/dvl-a50/dvl.py:
```diff
@@ -0,0 +1,234 @@
+        Load settings from .config/dvl/settings.json
+        """
+        try:
+            with open(os.path.join(os.path.expanduser("~"), ".config", "dvl", "settings.json")) as settings:
```
nice

Williangalvani commented on services/waterlinked/dvl-a50/dvl.py:
```diff
@@ -0,0 +1,234 @@
+                except Exception as e:
+                    print("Error receiveing:", e)
+            if not data:
+                time.sleep(0.003)
```
yep, we want to catch the next available data asap

Williangalvani replied:
the older(but newer) implemention with async was better in that regard, but we don't have that here =[

Williangalvani commented on services/waterlinked/dvl-a50/main.py:
```diff
@@ -0,0 +1,72 @@
+        return str(api.set_hostname(hostname))
+
+    dvl.start()
+    app.run(port=9001)
```
it is [this](https://github.com/bluerobotics/companion/blob/master/services/network/main.py#L173) + 1
I want to keep our services ports close together

patrickelectric replied:
Ok, great, I forgot the port that I was using!

Williangalvani commented on services/waterlinked/dvl-a50/mavlink2resthelper.py:
```diff
@@ -0,0 +1,125 @@
+
+    def __init__(self):
+        # store vision template data so we don't need to fetch it multiple times
+        self.vision_template = """{{"header":{{"system_id":255,"component_id":0,"sequence":0}},"message":{{"type":"VISION_POSITION_DELTA","time_usec":0,"time_delta_usec":{dt},"angle_delta":[{dRoll},{dPitch},{dYaw}],"position_delta":[{dx}, {dy}, {dz}],"confidence":{confidence}}}}}"""
```
oops, this is both a leftover of when mavlink2rest was not providing templates and me trying to reduce cpu usage my not generating the json every time...
I think I'll keep it though. but multi-lined

Williangalvani committed:
Create DVL service

Williangalvani commented on br-webui/index.js:
```diff
@@ -1665,6 +1669,15 @@ io.on('connection', function(socket) {
 	});
 	
+	socket.on('restart DVL driver', function(data) {
+		var cmd = child_process.exec('screen -X -S dvl-service quit', function(error, stdout, stderr) {
```
I want the "service manager" to be able to restart the service even if it goes into a bad state where it is not working properly and can't restart itself.

patrickelectric replied:
Can we somehow avoid this ?
Is it not possible to the service to do it itself ?
In the camera-manager, it 'part' is a thread that is restarted if something goes wrong, and the base code is manager for all threads and the interchange of data between them.

Williangalvani committed:
DVL: setup params automatically

Williangalvani committed:
DVL: do not feedback attitude to ekf

This seems to aggravate the gyro drift issue

Williangalvani committed:
mavlink2rest: update to 0.8

jaxxzer commented on params/dhcpd-server.conf:
```diff
@@ -4,6 +4,6 @@
 subnet 192.168.2.0 netmask 255.255.255.0 {
 	max-lease-time 60;
-	range 192.168.2.1;
+	range 192.168.2.10 192.168.2.20;
```
WHY?

Williangalvani replied:
When I got the dvl it had no support for static IP Address. Keeping it at 192.168.2.1 made it receive the same address I had on the topside, and made me spend a few hours trying to understand what was going on.

Williangalvani replied:
They do support static IP address now, though, so we can take this out

arendeutsch commented:
Is it possible to enable the dvl without using the web interface.
I'd like to enable it via a python script

Williangalvani commented:
> Is it possible to enable the dvl without using the web interface.
> I'd like to enable it via a python script

to enable:
GET /api/waterlinked-dvl/enable/true
to disable:
GET /api/waterlinked-dvl/enable/false

arendeutsch commented:
Thanks you very much.
I will test it today.

Just for reference, where would be the best place to share our experiences with dvl and waypoints testing/development and ask questions ?

I'm from Mørenot robotics in Norway

Williangalvani commented on services/waterlinked/dvl-a50/dvl.py:
```diff
@@ -0,0 +1,313 @@
+                continue
+            # got the right Ip
+            self.version = request(
+                "http://{0}/api/v1/about".format(self.hostname))
```
```suggestion
                "http://{0}/api/v1/about".format(ip))
```

wlkh commented on services/waterlinked/dvl-a50/dvl.py:
```diff
@@ -0,0 +1,313 @@
+import math
+import os
+
+HOSTNAME = "192.168.2.11"
```
```suggestion
HOSTNAME = "192.168.2.95"
```

Default static IP of the DVL

wlkh committed:
Fix pylint warning

wlkh committed:
Fix decode errors

socket.recv gives partial messages, so buffer them and use once a complete line has been received.

wlkh committed:
Scale confidence based Figure of merit

Williangalvani committed:
Update setup.py requirements

Williangalvani committed:
Dvl: use default DVL Ip address as default HOSTNAME

Williangalvani committed:
Dvl: use found ip address to check for dvl

Williangalvani committed:
webui: add /waterlinked to header

wlkh committed:
dvl: Detect TCP disconnection. Improve connection process.

wlkh commented:
My latest changes contains fixes to ensure TCP connection stability as well as setting the confidence in a more refined way.
Our testing shows that TCP disconnection is reliably detected and reconnected when available. I don't have anything to be added or fixed before this can be merged.

Williangalvani commented:
> My latest changes contains fixes to ensure TCP connection stability as well as setting the confidence in a more refined way.
> Our testing shows that TCP disconnection is reliably detected and reconnected when available. I don't have anything to be added or fixed before this can be merged.

This PR was actually pushed into [it's own branch in Companion](https://github.com/bluerobotics/companion/tree/dvl). I will likely close this PR to avoid confusion.
